### PR TITLE
Add zsh completition made by Holger Macht and improve

### PR DESCRIPTION
**Pull request by Thaodan**
Right now the zsh completition is packaged inside the zsh package on
OpenSUSE.
Several distributions either package it or the user has to import it
themselves.

It is better if it is stored here so any can pick it up in there zsh
package and put fixes in here.

### CODE REVIEW

File: contrib/osc.zsh [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+# 
+# Use the variables $ZSH_OSC_BUILD_TARGETS_EXTRA and $ZSH_OSC_PROJECTS_EXTRA to
+# extend the list of possible completions in your ~/.zshrc like that:
+#  export OSC_PROJECTS_EXTRA="Base:System Base:shells"
+#
+# version 0.2
+#
+
+OSC_BUILD_TARGETS="openSUSE_13.1 openSUSE_13.2 openSUSE_Tumbleweed openSUSE_Factory SLE_11_SP3 SLE_12"
+OSC_PROJECTS="openSUSE:Factory openSUSE:Tumbleweed openSUSE:13.2 openSUSE:13.1"
```
Comment (dmach) [inline contrib/osc.zsh:20]: I don't like the hard-coded lists.
On the other hand, I don't see any other reasonable alternative at the moment.
Could you update the values to the active repos and projects at least?

**Review by dmach — CHANGES_REQUESTED** [review]
Are you planning to maintain this completion file?
I'm not interested in maintaining it and I'm actually considering using argcomplete in the future (it's not a priority for the moment, that's why I'm postponing it).

Could you also extend `contrib/osc.spec` to get the completion installed so it works out of the box?

**Comment by Thaodan** [comment]
> Are you planning to maintain this completion file?

I can do so. So far it is maintained in the zsh package, I think here is the better place for doing so, more eyes can see the file.

**Comment by Thaodan** [comment]
> I'm not interested in maintaining it and I'm actually considering using argcomplete in the future (it's not a priority for the moment, that's why I'm postponing it).

Ah if that is possible the file could be replaced. However if I understand correctly a wrapper for these is still needed as argcomplete needs to be activated.

**Comment by Thaodan** [comment]
> Could you also extend `contrib/osc.spec` to get the completion installed so it works out of the box?

I can do that too. BTW why is the spec file in this repo but not the other package descriptions such as PKGBUILD?

**Comment by dmach** [comment]
>  BTW why is the spec file in this repo but not the other package descriptions such as PKGBUILD?

It's because we primarily package osc as rpm and that's also focus of some of the upstream tests.
There's PKGBUILD in [openSUSE:Tools/osc](https://build.opensuse.org/package/show/openSUSE:Tools/osc) for example.

File: contrib/osc.zsh (hunk 1 above)
Comment (dmach) [reply contrib/osc.zsh:20]: > Sure I can do that. The bash completition uses `~/.osc.projects` instead maybe something like this works but also the path is wrong as it is not inside `$XDG_CACHE_HOME`.

Yes, that needs to be fixed. I've submitted https://github.com/openSUSE/osc/pull/1364

Comment (Thaodan) [reply contrib/osc.zsh]: Projects are no longer depended on hardcoded list. Now on targets and arches are left.

Comment (Thaodan) [reply contrib/osc.zsh]: Now targets are done.

Comment (Thaodan) [reply contrib/osc.zsh]: Architectures are now fetched from the project repositories too. Should be all good now too.

**Comment by Thaodan** [comment]
The competition is now much closer to the bash/tcsh or in some areas further than that.

Some commands don't have competition for the switches they provide thou.

**Comment by dmach** [comment]
I've added a commit that updates spec with installing the completion.

I need more time to test this, because I'm still getting build targets such as `openSUSE_13.1` even when testing in a new container. I'm probably doing something wrong, just need to figure out what it is :)

**Comment by Thaodan** [comment]
> I've added a commit that updates spec with installing the completion.
> 
> I need more time to test this, because I'm still getting build targets such as `openSUSE_13.1` even when testing in a new container. I'm probably doing something wrong, just need to figure out what it is :)

any news?

File: contrib/osc.spec [hunk 2]
```diff
@@ -10,6 +10,7 @@
 %define completion_dir_bash %{_datadir}/bash-completion/completions
 %define completion_dir_csh %{_sysconfdir}/profile.d
 %define completion_dir_fish %{_datadir}/fish/vendor_completions.d
+%define completion_dir_zsh %{_sysconfdir}/zsh_completion.d
```
Comment (Thaodan) [inline contrib/osc.spec:13]: I don't think that's the correct path for completions.

**Review by dmach — APPROVED** [review]

**Pull request merged**